File: charter/gitpolicy.py
"""**Golden rule 0: one credential — PER FORGE.** Every git operation in a control plane —
and in every repo clone under it — authenticates with *that repo's own forge's* token over
HTTPS. No SSH keys, no 1Password agent, no commit signing. One credential per forge (not one
credential, full stop — a GitLab clone uses ``glab``'s token, a GitHub clone uses ``gh``'s),
so an agent never stalls on a key prompt, a signer hang, or a missing SSH config, on *any*
forge the control plane spans.

**SCOPE — the control plane and its clones, nothing outside.** Everything here is written
with ``git config --local``, so it applies to the control-plane repo and the clones under
``workspaces/``, and **never** to a developer's global/system git config. A developer's own
preferences (e.g. a global ``commit.gpgsign=true``) keep working everywhere else on their
machine; inside the control plane the local value simply wins, which is exactly what stops
an agent stalling on a signer prompt. Do **not** widen this to ``--global`` / an ``includeIf``
stanza — tests assert the boundary (`test_git_policy.py`).

This module makes that **mechanically true** rather than merely documented, by writing three
things into a repo's *local* git config (never global — we don't touch a developer's machine),
each derived from the ONE forge that repo actually talks to (:func:`forge_for`, resolved from
its ``origin`` remote):

1. ``credential.helper = !<cli> auth git-credential`` — git asks *that forge's* CLI for the
   token (:func:`policy_for`).
2. ``commit.gpgsign=false`` / ``tag.gpgsign=false`` — signing can't hang the agent, on any
   forge (:func:`policy_for`).
3. ``url.<https-base>.insteadOf`` for both of that forge's SSH forms — so even a repo whose
   *remote is an SSH URL* transparently transports over HTTPS+token (:func:`insteadof_for`).
   This is what makes a plain ``git push`` typed by any agent (or any persona sub-agent) work
   without SSH, whichever forge the repo lives on.

With (3) in place the rule holds even for clones the control plane didn't create; the
PreToolUse guard in :mod:`charter.hooks` then blocks the deliberate bypasses (explicit SSH
remotes, ``GIT_SSH_COMMAND``, ``-S``/``--gpg-sign``) — widened the same way, across every
forge host the control plane knows about (see ``hooks._known_forges``).
"""

from __future__ import annotations

from pathlib import Path

from . import config, util
from .forge.gitlab import GitLabForge


def policy_for(forge) -> dict[str, str]:
    """The `key = value` settings a repo on *forge* gets. One credential per forge: the
    forge's own CLI holds the token, and signing is off so no signer prompt can hang an
    agent."""
    return {
        "credential.helper": forge.credential_helper(),
        "commit.gpgsign": "false",
        "tag.gpgsign": "false",
    }


def insteadof_for(forge) -> tuple[str, tuple[str, ...]]:
    """`(https_base, ssh_forms)` so even a repo whose remote is an SSH URL transports
    over HTTPS with a token."""
    return forge.insteadof()


#: Back-compat defaults, kept as *derived* values (never a separately-maintained literal) —
#: GitLab was the only forge before this module went per-forge, and stays the fallback a
#: fresh/originless repo gets (see `forge_for`). Existing callers/tests that read these
#: module-level names keep working unchanged; `check`/`apply` themselves resolve per-repo.
POLICY: dict[str, str] = policy_for(GitLabForge())
HTTPS_BASE, SSH_FORMS = insteadof_for(GitLabForge())
_URL_KEY = f"url.{HTTPS_BASE}.insteadOf"


def _git(args, cwd):
    return util.run(["git", *args], cwd=cwd, check=False)


def is_git_repo(path: Path) -> bool:
    return (Path(path) / ".git").exists()


def _local_config(repo: Path) -> dict[str, list[str]]:
    """The repo's LOCAL config as ``{key: [values]}`` in **one** git call (git parses its own
    format, so this stays correct) — cheap enough to check every clone at preflight."""
    p = _git(["config", "--local", "--list", "-z"], cwd=repo)
    out: dict[str, list[str]] = {}
    if p.returncode != 0:
        return out
    for rec in (p.stdout or "").split("\0"):
        if not rec:
            continue
        key, _, val = rec.partition("\n")
        out.setdefault(key.strip().lower(), []).append(val)
    return out


def _get_all(repo: Path, key: str) -> list[str]:
    return _local_config(repo).get(key.lower(), [])


#: `check` returns exactly this (and only this) when a repo's origin host is neither a
#: default forge host nor one DECLARED in the active control plane's ``charter.toml`` —
#: genuinely unrecognised. Distinguished from ordinary drift (`key != value`) because it
#: can't be fixed by `apply`: there's no forge to apply a policy FOR. Before this existed,
#: an unrecognised host silently fell back to GitLab's policy, so `check` returned `[]`
#: (compliant) for a repo it had never actually verified — a false green that is worse
#: than no check at all, because it still LOOKS like the golden rule is enforced.
UNMANAGED_FORGE = ("forge unknown — origin host is not a default forge host and isn't "
                   "declared in charter.toml; the token-only policy can't be verified or "
                   "applied. Declare it under [[forge]] (host = \"…\") to bring it under "
                   "management.")


def forge_for(repo: Path) -> object | None:
    """Which forge governs *repo*'s policy — inferred from its ``origin`` remote.

    Recognises two things: each registered kind's DEFAULT host (``gitlab.com``,
    ``github.com``), and any self-hosted host the ACTIVE control plane DECLARES in its
    own ``charter.toml`` (``registry.resolve_host`` — the same host set the SSH guard
    uses, see ``hooks._known_forges``). This is what makes `check`/`apply` per-forge: a
    GitHub clone gets ``gh``'s credential helper and github.com's SSH→HTTPS rewrite, a
    self-hosted GitLab clone gets ``glab``'s and ITS OWN host's rewrite — never
    gitlab.com's, which would leave its real SSH remote unrewritten.

    Returns ``None`` when *repo* has an origin whose host is genuinely unrecognised —
    not a default, not declared. Callers MUST treat that as unmanaged, never silently
    fall back to another forge's policy (see `UNMANAGED_FORGE`); "do not simply widen
    the fallback" is the whole point — a wrong-but-plausible-looking policy is worse than
    an honest "can't tell".

    A repo with NO origin at all (a fresh ``git init``, nothing pushed yet) is a
    different case — there's no host to be wrong about — and keeps the pre-multi-forge
    default (GitLab), unchanged from before multi-forge support existed."""
    from .forge import registry
    url = _git(["remote", "get-url", "origin"], cwd=repo).stdout.strip()
    if not url:
        return GitLabForge()
    return registry.resolve_host(url, config.ROOT)


def check(repo: Path, cfg: dict[str, list[str]] | None = None) -> list[str]:
    """Policy settings that are MISSING/wrong in ``repo``'s local config (empty = compliant),
    against *that repo's own forge* (see `forge_for`). Pass ``cfg`` to reuse an
    already-read config (avoids a second git call).

    An unrecognised forge (`forge_for` → ``None``) returns ``[UNMANAGED_FORGE]`` —
    never ``[]``. Reporting compliant for a repo whose policy was never actually checked
    is the false-green bug this exists to prevent."""
    forge = forge_for(repo)
    if forge is None:
        return [UNMANAGED_FORGE]
    cfg = _local_config(repo) if cfg is None else cfg
    policy = policy_for(forge)
    https_base, ssh_forms = insteadof_for(forge)
    url_key = f"url.{https_base}.insteadOf"
    drift = []
    for key, want in policy.items():
        got = cfg.get(key.lower(), [])
        if not got or got[-1] != want:
            drift.append(f"{key} != {want}")
    have = set(cfg.get(url_key.lower(), []))
    for ssh in ssh_forms:
        if ssh not in have:
            drift.append(f"{url_key} missing {ssh}")
    return drift


def non_compliant(root: Path, workspaces_dir: Path) -> list[Path]:
    """Every repo in scope (control plane + clones) whose local config isn't token-only."""
    return [r for r in repos(root, workspaces_dir) if check(r)]


def apply(repo: Path) -> list[str]:
    """Write the token-only policy for *repo's own forge* into its **local** config
    (idempotent). Returns the settings changed. Never touches global/system config.

    A no-op (returns ``[]``) when the forge is unrecognised (`forge_for` → ``None``) —
    there's nothing safe to apply for a host we can't identify; guessing would just
    reintroduce the false-green bug one layer down."""
    repo = Path(repo)
    if not is_git_repo(repo):
        return []
    forge = forge_for(repo)
    if forge is None:
        return []
    policy = policy_for(forge)
    https_base, ssh_forms = insteadof_for(forge)
    url_key = f"url.{https_base}.insteadOf"
    changed = []
    for key, want in policy.items():
        got = _get_all(repo, key)
        if not got or got[-1] != want:
            _git(["config", "--local", key, want], cwd=repo)
            changed.append(f"{key}={want}")
    have = set(_get_all(repo, url_key))
    for ssh in ssh_forms:
        if ssh not in have:
            _git(["config", "--local", "--add", url_key, ssh], cwd=repo)
            changed.append(f"rewrite {ssh} → {https_base}")
    return changed


def repos(root: Path, workspaces_dir: Path) -> list[Path]:
    """The control plane itself plus every repo clone under ``workspaces/<ws>/<repo>``."""
    return scan(root, workspaces_dir)[0]


def scan(root: Path, workspaces_dir: Path) -> tuple[list[Path], list[Path]]:
    """:func:`repos`, and beside it every directory under ``workspaces/<ws>/`` whose ``.git``
    charter cannot check.

    #942 final review: `Path.exists` RAISED for one on 3.11–3.13 — a workspace's own `refs/` at
    mode 000 took `charter doctor` down — and answered False on 3.14, leaving what may be a clone
    out of the count without a word. `workspace._exists` tells those apart, and `doctor`'s
    `check_ssh` names the second kind (ADR 0009)."""
    from .workspace import _exists

    out = [Path(root)] if is_git_repo(root) else []
    unseen: list[Path] = []
    if Path(workspaces_dir).exists():
        for ws in sorted(Path(workspaces_dir).iterdir()):
            if not ws.is_dir():
                continue
            for clone in sorted(ws.iterdir()):
                there = _exists(clone / ".git", follow=True)
                if there:
                    out.append(clone)
                elif there is None:
                    unseen.append(clone)
    return out, unseen

File: crates/charter-core/src/gitpolicy.rs
//! Golden rule 0 in a clone's own config: one credential per forge, over HTTPS, no signing.
//!
//! A port of the half of `charter/gitpolicy.py` that `clone` runs — `apply` and what it
//! needs. It writes three things into a repo's **local** config, never the global one, each
//! derived from the one forge the repo's `origin` names:
//!
//! 1. `credential.helper = !<cli> auth git-credential` — git asks that forge's CLI;
//! 2. `commit.gpgsign = false`, `tag.gpgsign = false` — no signer prompt can hang an agent;
//! 3. `url.<https-base>.insteadOf` for both of the forge's SSH forms — so a plain `git push`
//!    typed later in that clone goes over HTTPS with the token even if its remote is SSH.
//!
//! What is written is exactly Python's, byte for byte in the clone's `.git/config`, so the
//! Python `charter git-policy` reads a Rust clone as compliant and the other way round.

use std::collections::BTreeMap;
use std::io;
use std::path::{Path, PathBuf};

use crate::doctor::fsx;
use crate::forge::{self, Forge, Kind};
use crate::repocmd::{Say, Sink};
use crate::worktree::git;

/// The forge that governs `repo`'s policy, from its `origin`; `None` for a host this plane
/// does not manage, which is never answered with another forge's policy. A repo with no
/// origin at all keeps the pre-multi-forge default, GitLab. Python's `forge_for`.
pub fn forge_for(repo: &Path, root: &Path) -> Option<Forge> {
    let url = git::run(repo, &["remote", "get-url", "origin"], git::READ)
        .map(|r| r.line().trim().to_string())
        .unwrap_or_default();
    if url.is_empty() {
        return Some(Forge::default_of(Kind::GitLab));
    }
    forge::resolve_host(&url, root)
}

/// The repo's local config as `{lowercased key: [values]}`, in one git call.
fn local_config(repo: &Path) -> BTreeMap<String, Vec<String>> {
    let mut out: BTreeMap<String, Vec<String>> = BTreeMap::new();
    let Ok(listed) = git::run(repo, &["config", "--local", "--list", "-z"], git::READ) else {
        return out;
    };
    if !listed.ok() {
        return out;
    }
    for record in listed.out.split('\0').filter(|r| !r.is_empty()) {
        let (key, value) = record.split_once('\n').unwrap_or((record, ""));
        out.entry(key.trim().to_lowercase())
            .or_default()
            .push(value.to_string());
    }
    out
}

/// Write the token-only policy for `repo`'s own forge into its local config, idempotently.
/// Returns what changed; nothing for a repo that is not one, or whose forge charter cannot
/// name. Python's `apply`.
pub fn apply(repo: &Path, root: &Path) -> Vec<String> {
    if std::fs::symlink_metadata(repo.join(".git")).is_err() {
        return Vec::new();
    }
    let Some(forge) = forge_for(repo, root) else {
        return Vec::new();
    };
    let policy = [
        ("credential.helper", forge.credential_helper()),
        ("commit.gpgsign", "false".to_string()),
        ("tag.gpgsign", "false".to_string()),
    ];
    let (https_base, ssh_forms) = forge.insteadof();
    let url_key = format!("url.{https_base}.insteadOf");
    let mut changed = Vec::new();
    for (key, want) in &policy {
        let got = local_config(repo)
            .remove(&key.to_lowercase())
            .unwrap_or_default();
        if got.last() != Some(want) {
            let _ = git::run(repo, &["config", "--local", key, want], git::READ);
            changed.push(format!("{key}={want}"));
        }
    }
    let have = local_config(repo)
        .remove(&url_key.to_lowercase())
        .unwrap_or_default();
    for ssh in &ssh_forms {
        if !have.contains(ssh) {
            let _ = git::run(
                repo,
                &["config", "--local", "--add", &url_key, ssh],
                git::READ,
            );
            changed.push(format!("rewrite {ssh} → {https_base}"));
        }
    }
    changed
}

// --------------------------------------------------------------------------------------- //
// the check half, and `charter git-policy`                                                  //
// --------------------------------------------------------------------------------------- //

/// What [`check`] returns — and the ONLY thing it returns — for a repo whose origin host is
/// neither a default forge host nor one DECLARED in this plane's `charter.toml`.
///
/// Distinguished from ordinary drift because it cannot be fixed by [`apply`]: there is no
/// forge to apply a policy FOR. Before this existed an unrecognised host silently fell back to
/// GitLab's policy, so `check` returned "compliant" for a repo it had never verified — a false
/// green, which is worse than no check at all because it still LOOKS like the golden rule is
/// enforced.
pub const UNMANAGED_FORGE: &str = "forge unknown — origin host is not a default forge host and \
                                   isn't declared in charter.toml; the token-only policy can't \
                                   be verified or applied. Declare it under [[forge]] (host = \
                                   \"…\") to bring it under management.";

/// Policy settings that are MISSING or wrong in `repo`'s local config, against THAT repo's own
/// forge. Empty means compliant. Python's `check`.
pub fn check(repo: &Path, root: &Path) -> Vec<String> {
    let Some(forge) = forge_for(repo, root) else {
        return vec![UNMANAGED_FORGE.to_string()];
    };
    let cfg = local_config(repo);
    let policy = [
        ("credential.helper", forge.credential_helper()),
        ("commit.gpgsign", "false".to_string()),
        ("tag.gpgsign", "false".to_string()),
    ];
    let (https_base, ssh_forms) = forge.insteadof();
    let url_key = format!("url.{https_base}.insteadOf");
    let mut drift = Vec::new();
    for (key, want) in &policy {
        let got = cfg.get(&key.to_lowercase());
        if got.and_then(|values| values.last()) != Some(want) {
            drift.push(format!("{key} != {want}"));
        }
    }
    let empty = Vec::new();
    let have = cfg.get(&url_key.to_lowercase()).unwrap_or(&empty);
    for ssh in &ssh_forms {
        if !have.contains(ssh) {
            drift.push(format!("{url_key} missing {ssh}"));
        }
    }
    drift
}

/// A directory under `workspaces/` whose `.git` charter could not look at, and the errno the
/// look met — which is what decides the remedy: a symlink loop is cleared at the loop, and
/// only a refusal by restoring read access. `doctor`'s `Unread`, by another name.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Unseen {
    pub path: PathBuf,
    pub code: Option<i32>,
}

/// Whether there is a `.git` at `path` — asked the way Python's `Path.exists` asks it, through
/// the link rather than at it.
fn is_git_repo(path: &Path) -> bool {
    std::fs::metadata(path.join(".git")).is_ok()
}

/// The control plane itself plus every repo clone under `workspaces/<ws>/<repo>`, and beside
/// them every directory whose `.git` charter cannot check. Python's `scan`.
///
/// **A level charter cannot LIST is one of them too.** `workspaces/` itself is then the only
/// entry, and one unreadable workspace is named beside the clones of every other — raising
/// there hid every other workspace's clones along with it, and took `charter doctor` down as a
/// whole command rather than one row.
pub fn scan(root: &Path, workspaces_dir: &Path) -> (Vec<PathBuf>, Vec<Unseen>) {
    let mut out = Vec::new();
    let mut unseen = Vec::new();
    if is_git_repo(root) {
        out.push(root.to_path_buf());
    }
    let workspaces = match sorted_dir(workspaces_dir) {
        Ok(entries) => entries,
        Err(e) if fsx::absent(&e) => return (out, unseen),
        Err(e) => {
            // The LISTING, guarded apart from each workspace below: nothing under it can be
            // counted, so it is named on its own and every clone in it is out of the count.
            return (
                out,
                vec![Unseen {
                    path: workspaces_dir.to_path_buf(),
                    code: e.raw_os_error(),
                }],
            );
        }
    };
    for ws in workspaces {
        let clones = match sorted_dir(&ws) {
            Ok(entries) => entries,
            // Not a directory — Finder's `.DS_Store`, or a link to nothing.
            Err(e) if fsx::absent(&e) => continue,
            Err(e) => {
                unseen.push(Unseen {
                    path: ws,
                    code: e.raw_os_error(),
                });
                continue;
            }
        };
        for clone in clones {
            match std::fs::metadata(clone.join(".git")) {
                Ok(_) => out.push(clone),
                Err(e) if fsx::absent(&e) => continue,
                Err(e) => unseen.push(Unseen {
                    path: clone,
                    code: e.raw_os_error(),
                }),
            }
        }
    }
    (out, unseen)
}

fn sorted_dir(dir: &Path) -> io::Result<Vec<PathBuf>> {
    let mut entries: Vec<PathBuf> = std::fs::read_dir(dir)?
        .filter_map(Result::ok)
        .map(|e| e.path())
        .collect();
    entries.sort();
    Ok(entries)
}

/// `charter git-policy [--apply]`. Python's `cmd_git_policy`. Always exits 0: drift is a report,
/// not a failure of the command that found it.
pub fn policy(root: &Path, apply_it: bool, say: Sink) -> u8 {
    let workspaces_dir = root.join("workspaces");
    let (targets, unseen) = scan(root, &workspaces_dir);
    // What the scan could not read, named FIRST and with what clears it. Reading the repo list
    // alone reported on the clones it reached and skipped the rest without a word, down to "No
    // git repos found".
    for missed in &unseen {
        let mut named = missed
            .path
            .strip_prefix(root)
            .unwrap_or(&missed.path)
            .to_string_lossy()
            .replace(std::path::MAIN_SEPARATOR, "/");
        if missed.path == workspaces_dir {
            named.push('/');
        }
        // The REMEDY is `doctor`'s, not a second copy of it: one `ELOOP` and one sentence, so
        // the `git auth` row and this command cannot send a reader to different repairs for
        // one directory. The path is named as Python names it here — absolute, in full.
        say(Say::Warn(format!(
            "{named} cannot be checked — {}",
            fsx::uncheckable_fix(missed.code, &missed.path.display().to_string())
        )));
    }
    if targets.is_empty() {
        say(Say::Info(
            "No git repos found (control plane + workspace clones).".into(),
        ));
        return 0;
    }
    let (mut drifted, mut fixed, mut unmanaged) = (0usize, 0usize, 0usize);
    for repo in &targets {
        let rel = repo.strip_prefix(root).unwrap_or(repo).to_string_lossy();
        let name = if rel.is_empty() || rel == "." {
            "control plane".to_string()
        } else {
            rel.into_owned()
        };
        let drift = check(repo, root);
        if drift.is_empty() {
            continue;
        }
        if drift.len() == 1 && drift[0] == UNMANAGED_FORGE {
            // Honest "can't tell" — never silently reported green, and never guessed at via
            // `apply` either.
            unmanaged += 1;
            say(Say::Warn(format!("{name}: {UNMANAGED_FORGE}")));
            continue;
        }
        drifted += 1;
        if apply_it {
            let changes = apply(repo, root);
            fixed += 1;
            say(Say::Done(format!(
                "{name}: applied {} setting(s) — token-only",
                changes.len()
            )));
        } else {
            say(Say::Warn(format!(
                "{name}: {} setting(s) not token-only",
                drift.len()
            )));
            for line in drift.iter().take(4) {
                say(Say::Info(format!("    {line}")));
            }
        }
    }
    if drifted == 0 && unmanaged == 0 {
        say(Say::Done(format!(
            "All {} repo(s) are token-only (each forge's own HTTPS token, no SSH, no signing).",
            targets.len()
        )));
        return 0;
    }
    if apply_it {
        say(Say::Done(format!(
            "Applied the single-credential policy to {fixed} of {} repo(s).",
            targets.len()
        )));
    } else if drifted > 0 {
        say(Say::Info(format!(
            "{drifted} of {} repo(s) drifted — fix: charter git-policy --apply",
            targets.len()
        )));
    }
    if unmanaged > 0 {
        say(Say::Warn(format!(
            "{unmanaged} repo(s) have an unrecognised forge — not covered by any policy. \
             Declare the host in charter.toml's [[forge]] to bring {} under management.",
            if unmanaged == 1 { "it" } else { "them" }
        )));
    }
    0
}

#[cfg(test)]
mod tests {
    use super::*;

    fn repo(origin: Option<&str>) -> (tempfile::TempDir, std::path::PathBuf) {
        let dir = tempfile::tempdir().unwrap();
        let root = std::fs::canonicalize(dir.path()).unwrap();
        std::fs::write(
            root.join("charter.toml"),
            "[[forge]]\nkind = \"gitlab\"\nhost = \"git.internal\"\n",
        )
        .unwrap();
        let clone = root.join("workspaces/alpha/widget");
        std::fs::create_dir_all(&clone).unwrap();
        // Deliberately NOT through `crate::testgit`: its template writes the very
        // `commit.gpgsign = false` these tests assert the policy writes, which would make them
        // pass without it. Nothing here commits, so no signer is ever asked.
        git::run(&clone, &["init", "-q", "."], git::READ).unwrap();
        if let Some(origin) = origin {
            git::run(&clone, &["remote", "add", "origin", origin], git::READ).unwrap();
        }
        (dir, clone)
    }

    fn config(repo: &Path) -> String {
        std::fs::read_to_string(repo.join(".git/config")).unwrap()
    }

    #[test]
    fn a_github_clone_gets_ghs_helper_no_signing_and_both_ssh_rewrites() {
        let (dir, clone) = repo(Some("https://github.com/acme/widget.git"));
        let root = std::fs::canonicalize(dir.path()).unwrap();

        let changed = apply(&clone, &root);

        assert_eq!(changed.len(), 5, "{changed:?}");
        let written = config(&clone);
        assert!(
            written.contains("helper = !gh auth git-credential"),
            "{written}"
        );
        assert!(written.contains("[commit]\n\tgpgsign = false"), "{written}");
        assert!(written.contains("[tag]\n\tgpgsign = false"), "{written}");
        assert!(
            written.contains(
                "[url \"https://github.com/\"]\n\tinsteadOf = git@github.com:\n\tinsteadOf = ssh://git@github.com/"
            ),
            "{written}"
        );
        assert!(
            apply(&clone, &root).is_empty(),
            "a second apply changes nothing"
        );
    }

    #[test]
    fn a_declared_self_hosted_forge_gets_its_own_host_never_the_default() {
        let (dir, clone) = repo(Some("git@git.internal:team/widget.git"));
        let root = std::fs::canonicalize(dir.path()).unwrap();

        apply(&clone, &root);

        let written = config(&clone);
        assert!(written.contains("!glab auth git-credential"), "{written}");
        assert!(
            written.contains("[url \"https://git.internal/\"]"),
            "{written}"
        );
        assert!(!written.contains("gitlab.com"), "{written}");
    }

    #[test]
    fn a_host_charter_does_not_manage_gets_no_policy_rather_than_another_forges() {
        let (dir, clone) = repo(Some("https://evil.example/acme/widget.git"));
        let root = std::fs::canonicalize(dir.path()).unwrap();

        assert!(apply(&clone, &root).is_empty());
        assert!(!config(&clone).contains("credential"), "{}", config(&clone));
    }
}
